source/common/network/io_socket_handle_impl.cc

```diff
@@ -107,14 +108,54 @@ Api::IoCallUint64Result IoSocketHandleImpl::sendmsg(const Buffer::RawSlice* slic
   message.msg_namelen = address_base->sockAddrLen();
   message.msg_iov = iov.begin();
   message.msg_iovlen = num_slices_to_write;
-  message.msg_control = nullptr;
-  message.msg_controllen = 0;
   message.msg_flags = 0;
-
   auto& os_syscalls = Api::OsSysCallsSingleton::get();
-  const Api::SysCallSizeResult result = os_syscalls.sendmsg(fd_, &message, flags);
-
-  return sysCallResultToIoCallResult(result);
+  if (self_ip == nullptr) {
+    message.msg_control = nullptr;
+    message.msg_controllen = 0;
+    const Api::SysCallSizeResult result = os_syscalls.sendmsg(fd_, &message, flags);
+    return sysCallResultToIoCallResult(result);
+  } else {
+#ifndef __APPLE__
+    constexpr int kSpaceForIpv4 = CMSG_SPACE(sizeof(in_pktinfo));
+    constexpr int kSpaceForIpv6 = CMSG_SPACE(sizeof(in6_pktinfo));
+    // kSpaceForIp should be big enough to hold both IPv4 and IPv6 packet info.
+    constexpr int kSpaceForIp = (kSpaceForIpv4 < kSpaceForIpv6) ? kSpaceForIpv6 : kSpaceForIpv4;
+    char cbuf[kSpaceForIp]{0};
+#else
+    // Currently CMSG_SPACE is not constexpr in MAC OS. 48 bytes should be
+    // enough for in_pktinfo or in6_pktinfo.
+    char cbuf[48]{0};
+#endif
+    message.msg_control = cbuf;
+    message.msg_controllen = sizeof(cbuf);
+    cmsghdr* const cmsg = CMSG_FIRSTHDR(&message);
+    RELEASE_ASSERT(cmsg != nullptr, fmt::format("cbuf with size {} is not enough, cmsghdr size {}",
+                                                sizeof(cbuf), sizeof(cmsghdr)));
+    if (self_ip->version() == Address::IpVersion::v4) {
+      cmsg->cmsg_level = IPPROTO_IP;
+#ifndef IP_SENDSRCADDR
```

Comment:
Could you add a short comment to explain which platforms have IP_SENDSRCADDR and which platforms have IP_PKTINFO?

Reply:
I see, thanks!